### PR TITLE
standalone PX4 

### PR DESCRIPTION
**Pull request by frede791**
Describes how to launch PX4 and Gazebo separately as per 

- [x] https://github.com/PX4/PX4-Autopilot/pull/22467.

### CODE REVIEW

**Comment by frede791** [comment]
@hamishwillee Please provide feedback on this.

File: en/sim_gazebo_gz/README.md
Comment (beniaminopozzan) [inline en/sim_gazebo_gz/README.md:180]: Remove this please, `gz_env_sh.in` has ben removed in PX4-Autopilot

**Review by beniaminopozzan — CHANGES_REQUESTED** [review]
Considering that if the PR get merged, Gazebo Classic and Gazebo will behave in a complete different way, in the sense that I can run `make sitl gazebo-classic` and the full simulation will start, while for Gazebo I need to perform one or multiple actions depending on my use cases:

- Add an additional env variable for having the simulation to start automatically.
- Manually specify and env variable to let Gazebo know where the models and worlds are.
- Clone a new repo and run scripts from it.

This should properly documented at the very beginning to avoid confusion.

File: en/sim_gazebo_gz/README.md
Comment (beniaminopozzan) [reply en/sim_gazebo_gz/README.md:180]: Wrong line matching, I meant

:::note
`gz_env.sh.in` is compiled and made available in `$PX4_DIR/build/px4_sitl_default/rootfs/gz_env.sh`
:::

**Comment by frede791** [comment]
@beniaminopozzan I don't really have a preference either way. @dagar prefers maintaining the old way as well with this being an optional aside, primarily because people have already been trained on using the `make` command. I don't really have a preference either way but if we keep it in the old way we would require additional downloads to get the models and worlds from the PX4-gazebo models repo.

**Comment by frede791** [comment]
@beniaminopozzan I have adapted the PR to maintain current functionality.

File: en/releases/main.md
Comment (hamishwillee) [inline en/releases/main.md:45]:
```suggestion
- [Gazebo] Separation of Gazebo and PX4 SITL.
  The two are now separated and can be launched independently.
  Gazebo retains all functionality it previously, but now also supports drag-and-drop via the resource spawner in Gazebo GUI.
  Added to PX4 in [PX4-Autopilot#22467](https://github.com/PX4/PX4-Autopilot/pull/22467).
```

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:50]: Why do we need to specify `gz_x500` if the make command is not going to launch it?

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:44]: Why would I ever use standalone mode? As a reader this looks a lot harder than just doing a make command as above.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:46]: So to be pedantic, prefixing the command with `GZ_PX4_STANDALONE` tells PX4 to start without also starting gz-server right? However it is still expecting GZ to start right? So mostly this is so you can have PX4 and GZ running in separate places? (again, why?)

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:61]: So by default it just starts with the grey plane world right? What vehicle does it start with by default? 

I think you're basically saying that this script is what feeds GZ with the environment variables it needs to choose the right model and world.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:68]: At no point have I been asked to download `PX4-gazebo-models` so this won't work. I am pretty sure this all gets copied into `/.simulation-gazebo` when you first run make with a gazebo target? Is this in the root of your current drive like it looks?

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:235]: How is this in my path?

Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md:235]: Do I need to enter  `--world default`?

File: en/sim_gazebo_gz/gazebo-models.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/gazebo-models.md:20]: Since I don't do anything specific to fetch or call this normally,  presumably PX4 calls this simulation-gazebo at least once?

File: en/sim_gazebo_gz/gazebo-models.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/gazebo-models.md:32]: These options are either/or, not ordered? If so, make them bullets.

File: en/sim_gazebo_gz/gazebo-models.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/gazebo-models.md:25]: What is gazebo resource spawner?
What is "fuel"?

File: en/sim_gazebo_gz/gazebo-models.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/gazebo-models.md:29]: Did we need to indicate standalone?

File: en/sim_gazebo_gz/gazebo-models.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/gazebo-models.md:24]: Do we need to indicate standalone?

File: en/sim_gazebo_gz/gazebo-models.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/gazebo-models.md:3]: It would be good to add a sentence after this, once I understand it, that says something along the lines of: "PX4 automatically fetches this information when needed (or similar)".

I can't do that because not clear when it is fetched.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:270]: So can you access worlds/vehicles still without adding them to the repo?

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md:235]: My understanding is that PX4 would get that it is using a particular vehicle from `PX4_SYS_AUTOSTART=4001` and Gazebo would determine the vehicle from  `PX4_SIM_MODEL=gz_x500`. IS that right?

**Comment by hamishwillee** [comment]
@frede791 I can kind of see how this works, but if we're going to make things harder we need to explain why it is useful. Sorry if my questions are repetitive.

Do we need to add a line to https://docs.px4.io/main/en/sim_gazebo_gz/vehicles.html about where vehicles come from?

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:50]: The `make` command will launch PX4, it is just waiting in a holding loop, trying to connect to an instance of Gazebo Sim. Once an instance of gz-server, then PX4 will connect as normal.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:44]: Because you can launch Gazebo and PX4 across any two devices on your network and they will still connect. Up to now, PX4 and Gazebo always had to be on the same host. Now you can launch PX4 in a docker container or on a different PC across the building and Gazebo will still be able to connect. Furthermore, if you use a VPN, you can even connect the two from different networks.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:46]: Yes. It will start PX4 without starting gz-server. You don't have to start GZ, but PX4 will not be usable in that case. It will just print the warning message over and over again.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:61]: Grey plane is always default. There is no vehicle default (same way there is nothing to see when you run just `make px4_sitl`). This script will just launch the world with the necessary environmental variables (namely `GZ_PARTITION` and `GZ_IP`) and PX4 will then connect and provide the name of the desired vehicle. gz-server will then go look for that model in its own resource paths (found in `/.simulation-gazebo` per default) and load it into the world. 
Additional point that is not mentioned at the moment: If you want to run PX4 and GZ across different hosts, you will need to set `GZ_PARTITION` and `GZ_IP` on the PX4 host, as well as `GZ_RELAY`. At the moment, this is not mentioned on the user-guide but it should probably be mentioned, because this can be tricky to work out on your own. You can find an overview of the variables and their definitions here: https://gazebosim.org/api/transport/12/envvars.html

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:68]: I make a refernce to it here : `The simplest way to start the simulation is to run the `simulation-gazebo` script found in the [PX4-gazebo-models](https://github.com/PX4/PX4-gazebo-models) repository.` 

The thing is that you don't need to download the whole repository. You only really need this one file. Per default it is installed in root at `/.simulation-gazebo`. But you can of course provide a different path. But then you always need to set that path when you run the script.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:235]: It is not your path. I will adjust this. You do not actually need to enter `--world default`. However, I thought it would be useful to show in an example. I can take it out if you want. 
On your third question: Depends. If you start PX4 with `make` then you only need to provide the vehicle and PX4 would infer the id from that (and gazebo would receive the name). If you start with the binary directly and set the env variables, then your description is correct.

File: en/sim_gazebo_gz/gazebo-models.md
Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md:20]: Not the script itself. But it goes through some of the same motions. It checks for the existence of a folder called `/.simulation-gazebo` and downloads it from github if it doesn't exist or the `PX4_GZ_OVERWRITE` parameter has been set. It is set here: https://github.com/PX4/PX4-Autopilot/blob/d8807ac894ff7f8c139c9f1691c2a713f84a0029/ROMFS/px4fmu_common/init.d-posix/px4-rc.simulator#L75

Note that you don't get the same kind of freedoms when you run it this way (no custom storage directory option, no custom download directory option). Plus you also don't get to set the GZ env variables as mentioned above, so you cannot run PX4 and GZ separately.

File: en/sim_gazebo_gz/gazebo-models.md
Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md:32]: Ok

File: en/sim_gazebo_gz/gazebo-models.md
Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md:29]: No

File: en/sim_gazebo_gz/gazebo-models.md
Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md:24]: Yes. I will add standalone.

File: en/sim_gazebo_gz/gazebo-models.md
Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md:3]: Will add this.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:270]: For worlds, you only need to add the name of your world to the `worlds_list` file. For vehicles, you only need to add the airframe file to PX4 (and add that file to CMake) and then place the respective sdf files in a place where they lie on the GZ_SIM_RESOURCE_PATH. Per default this is in `/.simulation-gazebo` but you can modify this.

**Comment by frede791** [comment]
@hamishwillee This would be useful. 
I will also make the necessary changes to the points you raised. Let me know what you think about the remaining open questions.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md:68]: This is worth saying. Done.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md]: > You do not actually need to enter --world default. However, I thought it would be useful to show in an example. I can take it out if you want.

I would probably have kept this, or shown both :-)

Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md]: But no need to change it back.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md:288]: I think this is worth saying (right)? Can you adjust the docs at this point if you think they need any work in this respect?

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [reply en/sim_gazebo_gz/README.md]: >  If you want to run PX4 and GZ across different hosts, you will need to set GZ_PARTITION and GZ_IP on the PX4 host, as well as GZ_RELAY. At the moment, this is not mentioned on the user-guide but it should probably be mentioned, because this can be tricky to work out on your own. You can find an overview of the variables and their definitions here: https://gazebosim.org/api/transport/12/envvars.html

I think it would be useful to provide some examples of this in the "Gazebo Models Repository" doc. Leaving to you to do.

**Comment by hamishwillee** [comment]
@frede791 Excellent additions and thanks for the explanations. I have given this a subedit, primarily for structure now that I know how the bits fit together.

1. The standalone mode is its own section BELOW the vehicle list. This gives us a heading to link to in order to explain what it is. It is also much less confusing to read this when the modes are separate.
2. I link to the Gazebo Models Repository _document_ rather than repo generally. This ensures that people get an explanation of what it is used for before jumping to it. 
3. I've made it a bit more clear that you can just fetch the script any way you like. 

I've closed the discussions I think you have addressed. The ones that are open, such as https://github.com/PX4/PX4-user_guide/pull/2965#discussion_r1426180440 are those where it looked like you thought we might need more information. I'm hoping you will add that.

@beniaminopozzan Before this goes in would be good to have another review from you. From my perspective this now passes the "Hamish test" - I think I understand what this is for and could use it as described in the document. My only concern with that would be using across a network, for which no example is yet provided.

**Comment by frede791** [comment]
@hamishwillee: @dagar and I discussed last night that PX4-gazebo-models will remain included in PX4 as a submodule. I will make the adjustments accordingly.

**Comment by frede791** [comment]
@hamishwillee: I've amended the PR where necessary. In essence there now two locations where models and worlds are stored. 
If you run in "normal" mode, the worlds and models are taken from the submodule directly.
If you run in "standalone: mode, the simulation-gazebo script will fetch models and worlds from Github and place them in `~/.simulation-gazebo`. 

I will see if I can provide an example for the networking case.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md]: I've added a networking example that explains how to connect PX4 and gazebo on a single host and an example on how to run it on multiple hosts.

File: en/sim_gazebo_gz/README.md
Comment (frede791) [reply en/sim_gazebo_gz/README.md:297]: This is outdated now that the worlds list no longer exist. I've reworked the above mentioned section to explain the fact that we are now using a submodule and a different CMake structure.

File: en/sim_gazebo_gz/gazebo-models.md
Comment (beniaminopozzan) [inline en/sim_gazebo_gz/gazebo-models.md:11]: Is the repetition with the sections below intended?

**Review by beniaminopozzan — COMMENTED** [review]
LGTM!

Thanks @frede791 !

File: en/sim_gazebo_gz/gazebo-models.md
Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md]: @beniaminopozzan I am not quite sure what you are refering to here?

Comment (beniaminopozzan) [reply en/sim_gazebo_gz/gazebo-models.md]: I mean that
> For standalone builds you first have obtain the [simulation-gazebo](https://github.com/PX4/PX4-gazebo-models/blob/main/simulation-gazebo) script, and then it will fetch the models and worlds to `~/.simulation-gazebo` if that directory is not present.
The build system won't automatically update the local copy again if the `.simulation-gazebo` folder is detected, but you can force it to update to the latest models and vehicles by passing the `overwrite` flag to the script. The resulting command will look something like:

is quite similar to
>This will fetch the models and worlds from the [PX4 gazebo models repository](https://github.com/PX4/PX4-gazebo-models) into subfolders of the `~/.simulation-gazebo` directory the first time it is called (or more precisely, if the directory is not detected).
A _gz-server_ instance will then be launched using the default grey plane world.
The build system won't automatically update the local copy again if the `.simulation-gazebo` folder is detected, but you can force it to update to the latest models and vehicles by passing the `overwrite` flag to the script. The resulting command will look something like:

But nothing serius, just mentioning it.

Comment (frede791) [reply en/sim_gazebo_gz/gazebo-models.md]: No this is valid. the second sentence is identical. I don't know how this happened but I will fix it.

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:269]: Below is pretty good, but probably not quite enough to understand how to create your own airframe file for Gazebo.

Specifically, there are other things in the file you link to such as `param set-default SIM_GZ_EC_FUNC1` and these aren't mentioned. 

It would be good to have an understanding of end to end how you create a frame for posix, answering questions like "can I have a single frame for both posix and nuttx?"

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:289]: The rest of this is about how you add models to PX4. The assumption being that after this, someone else who wants to follow the instructions will be able to use the models themselves after downloading a version with this airframe and CMAKEList change after the models repo has been updated.

However, then we have these bullet points that say you have to copy the files to particular locations. You only need to do this when you're testing. 

So I am not going to wait on this, but do you think it makes sense to say something like. Once the models are updated to the repo they will be automatically to the right location when you install PX4. When testing you can copy them manually .... then the bullets ?

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:297]: Same comment as one above - this slightly feels like it is mixing "adding a world for everyone" and "adding a world for local testing".

File: en/sim_gazebo_gz/README.md
Comment (hamishwillee) [inline en/sim_gazebo_gz/README.md:304]: Again, mixing "Adding for everyone" and "adding for local testing".

I think though this is saying that as long as you put the model/world in those directories they will be found using the normal build syntax?

**Comment by hamishwillee** [comment]
@frede791 Thanks very much for this. I've subedited and I think "good enough to go in" - so will merge.

I had a couple of clarifications/questions around the adding worlds/models. I think they slightly mix what you do to get your world as part of the repo, and what you do to get it locally. I think both are useful things to cover, but it is important not to mix them in such a way it isn't obvious what you need to do for both cases. (I mean, it kind of is, but not everyone will necessarily have the English to untangle it).

**Comment by frede791** [comment]
@hamishwillee I will work on updating this.

**Comment by hamishwillee** [comment]
THanks very much @frede791 . I'm going to be in and out for the next month, but if you are able to look at this before Wednesday I will make sure it merges.

**Pull request merged**
